Local Network Access: Implement the Secure Context Restriction

rreno opened the pull request:
#### 4694c16e36d83d3efd77cc0091d128b0bb319722
<pre>
Local Network Access: Implement the Secure Context Restriction
<a href="https://bugs.webkit.org/show_bug.cgi?id=250330">https://bugs.webkit.org/show_bug.cgi?id=250330</a>
rdar://103142696

Reviewed by NOBODY (OOPS!).

Local Network Access has a Secure Context Restriction [0] which blocks a page in
the public address space that is a non-secure context from accessing resources
in the private or local address spaces. This adds a check and associated helpers
for checking that the request is allowed.

[0]: <a href="https://wicg.github.io/local-network-access/#secure-context-restriction">https://wicg.github.io/local-network-access/#secure-context-restriction</a>

* Source/WTF/wtf/text/StringToIntegerConversion.h:
(WTF::parseInteger):
(WTF::parseIntegerAllowingTrailingJunk):
(WTF::parseIntegerDisallowLeadingAndTrailingSpaces):
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake
* Source/WebCore/Modules/fetch/FetchLoader.cpp:
(WebCore::FetchLoader::start):
* Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.cpp: Added.
(WebCore::isPrivateNetworkAccessSecureContextRestricted):
* Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/SecurityContext.h:
(WebCore::SecurityContext::addressSpace const):
(WebCore::SecurityContext::setAddressSpace):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::prepareFetch):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::setAddressSpace):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::isValidIP4Address):
(WebCore::isLoopbackIPAddress):
(WebCore::SecurityOrigin::isPrivateIPv4Address):
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::addressSpace const):
(WebCore::ResourceRequestBase::setAddressSpace):
* Tools/TestWebKitAPI/Tests/WTF/StringToIntegerConversion.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4694c16e36d83d3efd77cc0091d128b0bb319722

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36773 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113048 "Hash 4694c16e for PR 8407 does not build (failure)") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173358 "Hash 4694c16e for PR 8407 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3834 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96071 "Hash 4694c16e for PR 8407 does not build (failure)") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112159 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10767 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93842 "Found 1 new API test failure: TestWebKitAPI.NetworkProcess.CORSPreflightCachePartitioned (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96071 "Hash 4694c16e for PR 8407 does not build (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92598 "Found 1 new API test failure: TestWebKitAPI.NetworkProcess.CORSPreflightCachePartitioned (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25450 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96071 "Hash 4694c16e for PR 8407 does not build (failure)") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/93935 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6295 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26835 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/90391 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4063 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3366 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29846 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12447 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46360 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/99000 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8231 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/24916 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.cpp:
```diff
@@ -0,0 +1,48 @@
+    if (securityContext.securityOrigin() && securityContext.securityOrigin()->isLocal())
+        return false;
+
+    // FIXME: add support for checking for IPv6 addresses in private address space.
```
Please file a Bugzilla (e.g., 123456) for this, and use `// FIXME(123456):` for the comment.

brentfulgham commented on Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.h:
```diff
@@ -0,0 +1,62 @@
+    Local
+};
+
+bool isPrivateNetworkAccessSecureContextRestricted(const SecurityContext& securityContext, const URL& url);
```
No need for '`securityContext`' label for the argument in the header.

brentfulgham reviewed (commented):
If this work implements any part of the Private Network Access specification, I'd encourage you to update `WebCore/features.json` to reflect that this is now in 'Development'.

rreno committed:
Local Network Access: Implement the Secure Context Restriction
https://bugs.webkit.org/show_bug.cgi?id=250330
rdar://103142696

Reviewed by NOBODY (OOPS!).

Local Network Access has a Secure Context Restriction [0] which blocks a page in
the public address space that is a non-secure context from accessing resources
in the private or local address spaces. This adds a check and associated helpers
for checking that the request is allowed.

[0]: https://wicg.github.io/local-network-access/#secure-context-restriction

* Source/WTF/wtf/text/StringToIntegerConversion.h:
(WTF::parseInteger):
(WTF::parseIntegerAllowingTrailingJunk):
(WTF::parseIntegerDisallowLeadingAndTrailingSpaces):
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake
* Source/WebCore/Modules/fetch/FetchLoader.cpp:
(WebCore::FetchLoader::start):
* Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.cpp: Added.
(WebCore::isPrivateNetworkAccessSecureContextRestricted):
* Source/WebCore/Modules/private-network-access/PrivateNetworkAccess.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/SecurityContext.h:
(WebCore::SecurityContext::addressSpace const):
(WebCore::SecurityContext::setAddressSpace):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::prepareFetch):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::setAddressSpace):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::isValidIP4Address):
(WebCore::isLoopbackIPAddress):
(WebCore::SecurityOrigin::isPrivateIPv4Address):
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::addressSpace const):
(WebCore::ResourceRequestBase::setAddressSpace):
* Tools/TestWebKitAPI/Tests/WTF/StringToIntegerConversion.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST):
